experimental/inductor.py

```diff
@@ -0,0 +1,218 @@
+from functorch.compile import min_cut_rematerialization_partition
+
+from torch._functorch.eager_transforms import functionalize
+
+from torch import autograd
+from torch._inductor.compile_fx import compile_fx_inner
+
+
+from torch._inductor.decomposition import select_decomp_table
+inductor_decomps = select_decomp_table()
+
+
+
+from copy import copy
+all_decomps = copy(inductor_decomps)
+
+
+def prepare_args(attr_vals, unused_inputs_idx, args):
+    """
+    modify args to pass into a function compiled with make_functional_fx
+    args: original arguments
+    unused_inputs_idx, attr_vals: data returned from make_functional_fx
+    """
+    all_args = list(args) + list(attr_vals)
+    return [a for i, a in enumerate(all_args) if i not in unused_inputs_idx]
+
+
+def make_functional_fx(fn):
+    """
+    Given a function, does make_fx(functionalize(make_fx(fn))).
+    Some input-related transformations need to be done before functionalize.
+    Returns an GraphModule to be fed to inductor.
+    """
+    def call(*args, **kwargs):
+        # 1. call make_fx for the first time
+        k1 = make_fx(fn)
+        k2 = k1(*args, **kwargs)
+
+
+        # 2. Turn getattr into inputs
+        #    - This is needed for functionalize to work. Functionalize expects
+        #      all tensors to be passed as inputs to the function being functionalized.
+        #    - We collect all concrete getattr inputs into "attr_vals"
+        from collections import defaultdict
+        attr_vals_map = dict()
+        attr_nodes = defaultdict(lambda: [])
+
+        for p in k2.graph.nodes:
+            if p.op == 'get_attr':
+                val = getattr(k2, p.target)
+                attr_vals_map[p.target] = val
+                attr_nodes[p.target].append(p)
+
+        after_last_placeholder = None
+        for i, p in enumerate(k2.graph.nodes):
+            if p.op != 'placeholder':
+                after_last_placeholder = p
+                break
+
+        attr_vals = []
+        with k2.graph.inserting_before(after_last_placeholder):
+            for attr, nodes in attr_nodes.items():
+                placeholder = k2.graph.placeholder(attr)
+                attr_vals.append(attr_vals_map[attr])
+                for node in nodes:
+                    node.replace_all_uses_with(placeholder)
+                    k2.graph.erase_node(node)
+    
+
+        # 3. Remove unused inputs
+        #    - This is needed in order to remove non-tensor inputs before passing this
+        #      graph to inductor. Inductor expects only tensors as inputs
+        #    - This works since all the non-tensor inputs should be unused at this point,
+        #      since in principle we should have converted all getattrs into inputs.
+        #    - We keep the indices of unused inputs in "unused_inputs_idx"
+        unused_inputs = []
+        unused_inputs_idx = []
+
+        for i, p in enumerate(k2.graph.nodes):
+            if p.op != 'placeholder':
+                break
```

Comment:
why this constraint? if you have unused other ops, might as well prune no?

Reply:
right and I do call eliminate_dead_code() at the end but it's a bit more complicated because there's some blacklisting that needs to happen and in particular there is an inplace call to copy_ that needs to be kept even though there's no user to its output.